src/backend/columnar/columnar_tableam.c

```diff
@@ -296,6 +297,22 @@ columnar_getnextslot(TableScanDesc sscan, ScanDirection direction, TupleTableSlo
  */
 static ItemPointerData
 row_number_to_tid(uint64 rowNumber)
+{
+	ErrorIfInvalidRowNumber(rowNumber);
+
+	ItemPointerData tid = { 0 };
+	ItemPointerSetBlockNumber(&tid, rowNumber / VALID_ITEMPOINTER_OFFSETS);
+	ItemPointerSetOffsetNumber(&tid, rowNumber % VALID_ITEMPOINTER_OFFSETS +
+							   FirstOffsetNumber);
+	return tid;
+}
+
+
+/*
+ * ErrorIfInvalidRowNumber errors out if given rowNumber is invalid.
+ */
+static void
+ErrorIfInvalidRowNumber(uint64 rowNumber)
```

Comment:
Don't we also need to check for very large row numbers that can't fit in the TID?

Reply:
I guess this comment was outdated, now `ErrorIfInvalidRowNumber` checks for `COLUMNAR_MAX_ROW_NUMBER` too